Development/nmos/jwk_utils.cpp

```diff
@@ -277,111 +217,186 @@ namespace nmos
                     { U("use"), pubkey_use.name }
                 });
             }
+#else
+                BIGNUM* modulus = nullptr;
+                BIGNUM* exponent = nullptr;
 
-            // convert RSA public key to JSON Web Key
-            web::json::value public_key_to_jwk(const utility::string_t& pubkey, const utility::string_t& keyid, const jwk::public_key_use& pubkey_use, const jwk::algorithm& alg)
-            {
-                using ssl::experimental::BIO_ptr;
-
-                const std::string public_key{ utility::us2s(pubkey) };
-                BIO_ptr bio(BIO_new_mem_buf((void*)public_key.c_str(), (int)public_key.length()), &BIO_free);
-                if (!bio)
+                utility::string_t base64_n;
+                if (EVP_PKEY_get_bn_param(pkey.get(), OSSL_PKEY_PARAM_RSA_N, &modulus))
                 {
-                    throw jwk_exception("convert pem to jwk error: failed to create BIO memory from public key");
+                    base64_n = to_base64url(modulus);
+                    BN_clear_free(modulus);
                 }
 
-                RSA* rsa_ = NULL;
-                RSA_ptr rsa(PEM_read_bio_RSA_PUBKEY(bio.get(), &rsa_, NULL, NULL), &RSA_free);
-                if(!rsa)
+                utility::string_t base64_e;
+                if (EVP_PKEY_get_bn_param(pkey.get(), OSSL_PKEY_PARAM_RSA_E, &exponent))
                 {
-                    throw jwk_exception("convert pem to jwk error: failed to load RSA");
+                    base64_e = to_base64url(exponent);
+                    BN_clear_free(exponent);
                 }
 
-                return rsa_to_jwk(rsa, keyid, pubkey_use, alg);
+                // construct jwk
+                return web::json::value_of({
+                    { U("kid"), keyid },
+                    { U("kty"), U("RSA") },
+                    { U("n"), base64_n },
+                    { U("e"), base64_e },
+                    { U("alg"), alg.name },
+                    { U("use"), pubkey_use.name }
+                });
             }
+#endif
+        }
```

Comment:
```suggestion
#endif
            }
        }
```

Reply:
Actually, would it be OK... nice... to move the `construct jwk` return-statement outside the #if-#else-#endif as well?

Reply:
yes, sure

Reply:
I am happy to change it.